### PR TITLE
Feature checkout, Reviews, Booking

### PR DESCRIPTION
**Pull request by CamiloGilFranco**
## SUMMARY 

- Verification of login and existence of products in the shopping cart is made

- The address list is filled according to the logged in user

- Function is given to the remove address button, eliminating the data from the database and rendering the list of addresses again

- Functionality is given to the new address form, adding it to the database and rendering the address list again

- Functionality is given to the newReview field restricting the publication of a new comment if the user is not logged in or if he has already made a new comment

- When making a new comment, the page is rendered again to show the new rating of the restaurant and the user's comment

- The form to create a reservation is modified and validations are applied

- Restriction is applied to the creation of a reservation by non-logged in users

- The reservation creation functionality is completed

- Stripe payment gateway is implemented

- The test payment form is removed and the native Stripe is implemented

- When making payment, you are redirected to the successful payment page

- Add id_dish and id_restaurant to the global state of cartItem

- Once the payment is confirmed, a new order record is added to the database with its corresponding order_details

- Once the purchase is completed, the customer is redirected to the main page

### CODE REVIEW

File: src/components/CartItem/CartItem.jsx [hunk 1]
```diff
@@ -37,15 +65,17 @@ const CartItem = () => {
             ${cartSubtotal}
           </span>
         </div>
-        <p className="cart-item-finish-waring">{languageSelector(language, 'cartFinishWarning')}</p>
+        <p className="cart-item-finish-waring">
```
Comment (ht1204) [inline src/components/CartItem/CartItem.jsx]: `cart-item-finish-warning`

File: src/components/DeliveryAddressComponent/DeliveryAddressComponent.jsx [hunk 2]
```diff
@@ -24,6 +45,8 @@ const DeliveryAddressComponent = () => {
     setAddressList([...addressList, newAddress]);
   };
 
+  console.log();
```
Comment (ht1204) [inline src/components/DeliveryAddressComponent/DeliveryAddressComponent.jsx]: what is this console.log doing?

File: src/components/DeliveryAddressComponent/DeliveryAddressComponent.jsx [hunk 3]
```diff
@@ -39,25 +62,31 @@ const DeliveryAddressComponent = () => {
...
-              mobileNumber={element.mobileNumber}
-              address={element.address}
-              city={element.city}
-              setAddressSelected={setAddressSelected}
-              addressSelected={addressSelected}
-              handleDelete={handleDelete}
-            />
-          ))}
+          {addressesList &&
+            addressesList.map((element, index) => (
```
Comment (ht1204) [inline src/components/DeliveryAddressComponent/DeliveryAddressComponent.jsx]: After discussing with Sergio about this, we do recommend this conditional rendering for mapping:
```js
{!!addressesList.map((element, index) => ( ...
```

File: src/components/OrderSuccesful/OrderSuccesful.jsx [hunk 4]
```diff
@@ -1,24 +1,100 @@
...
+            dishesId_dish: dish.id,
+            restaurantsId_restaurant: dish.id_restaurant,
+          });
+        }
+        axios
+          .post(`${API_URL}/orders-details/several`, {
+            dishList,
+          })
+          .then(() => {
+            toast.success("Going back to the main page", {
```
Comment (ht1204) [inline src/components/OrderSuccesful/OrderSuccesful.jsx]: I consider that toast messages should be better in constants for importing when are necessaries.

File: src/components/ReserveForm/ReserveForm.jsx [hunk 5]
```diff
@@ -1,161 +1,158 @@
...
+            `${API_URL}/reservations`,
+            { date: reserveDate, id_restaurant, id_venue: reserveForm.venue },
+            {
+              headers: {
+                Authorization: `Bearer ${cookies.get("token")}`,
+              },
+            }
+          )
+          .then(() => {
+            toast.success("your reservation was made successfully", {
```
Comment (ht1204) [inline src/components/ReserveForm/ReserveForm.jsx]: Texts start with upper case: `"Your..."`

File: src/components/ReserveForm/ReserveForm.jsx [hunk 6]
```diff
@@ -1,161 +1,158 @@
...
+              position: "bottom-right",
+            });
+            setReserveForm({ dateAndTime: "", venue: "CHOOSE ONE" });
+          })
+          .catch(() => {
+            toast.error(
+              "your reservation was not completed, try again or select another date",
```
Comment (ht1204) [inline src/components/ReserveForm/ReserveForm.jsx]: here

File: src/components/ReserveForm/ReserveForm.jsx [hunk 7]
```diff
@@ -1,161 +1,158 @@
...
+                });
+              }}
+            />
+          </div>
+          {dateError ? (
+            <p className="reserve-form-input-error-message">
+              * debes reservar una fecha menor a 30 días
+            </p>
+          ) : (
+            ""
```
Comment (ht1204) [inline src/components/ReserveForm/ReserveForm.jsx]: `<></>`

File: src/components/ReserveForm/ReserveForm.jsx [hunk 8]
```diff
@@ -1,161 +1,158 @@
...
+                ...reserveForm,
+                venue: event.target.value,
+              })
+            }
+            value={reserveForm.venue}
+            className="reserve-form-input"
+          >
+            <option value="CHOOSE ONE">CHOOSE A VENUE</option>
+            {venues &&
+              venues.map((venue, index) => {
```
Comment (ht1204) [inline src/components/ReserveForm/ReserveForm.jsx]: `!!venues.map`

File: src/components/RestaurantList/RestaurantList.jsx [hunk 9]
```diff
@@ -126,18 +127,22 @@ const RestaurantList = () => {
...
-            rating={element.rating}
-            categories={element.cuisines}
-            path={element.restaurant_path}
-          />
-        );
-      });
+      return (
+        !!listToRender &&
+        listToRender.length &&
+        listToRender.map((element) => {
```
Comment (ht1204) [inline src/components/RestaurantList/RestaurantList.jsx]: `!!listToRender.map`

File: src/components/carrusel/CarrucelComponent.jsx [hunk 10]
```diff
@@ -59,31 +60,33 @@ const CarrucelComponent = () => {
...
-              >
-                <div className="home-carousel-single-item-image-container">
-                  <img
-                    src={element.cuisine_photo}
-                    alt=""
-                    className="home-carousel-single-item-image"
-                  ></img>
+          {!!cuisinesList &&
+            cuisinesList.length > 0 &&
+            cuisinesList.map((element, index) => {
```
Comment (ht1204) [inline src/components/carrusel/CarrucelComponent.jsx]: Only this:
`!!cuisinesList.map((element, index)`

File: src/components/payment/payment.css [hunk 11]
```diff
@@ -44,13 +44,22 @@
 }
 
 .debit-card-text-input {
-  font-size: 20px;
+  font-size: 20px !important;
```
Comment (ht1204) [inline src/components/payment/payment.css:47]: Hope there is another way for font styling.

File: src/pages/CheckoutPageComponent/CheckoutPageComponent.jsx [hunk 12]
```diff
@@ -5,8 +5,19 @@ import Footer from "../../components/Footer/Footer";
 import image from "./assets/pexels-ella-olsson-1640777.jpg";
 import { Payment } from "../../components/payment/Payment";
 import "./CheckoutPageComponent.css";
+import { Elements } from "@stripe/react-stripe-js";
+import { loadStripe } from "@stripe/stripe-js";
+import { useState } from "react";
+
+const stripePromise = loadStripe(
+  "pk_test_51MzwoXA4m6OMz50cQDWoJL27hwzyeiodGS1ivBKyHMNGu0d250XsB9IhkbB89Ig8kIoN0Lkq8PTos0bG9F5YY2gr00jbNcg0q0"
```
Comment (ht1204) [inline src/pages/CheckoutPageComponent/CheckoutPageComponent.jsx]: This API Key would be better in a constant

File: src/pages/RestaurantAdminView/RestaurantAdminView.jsx [hunk 13]
```diff
@@ -11,7 +22,37 @@ import EditDetailsModal from '../../components/EditDetailsModal/EditDetailsModal
...
+      Authorization: `Bearer ${jwtToken}`,
+    },
+  };
+  
+  useEffect(() => {
+    const fetchData = async () => {
+      dispatch(fetchUserRequest());
+      try {
+        const userResponse = axios.get(`${API_URL}/users/profile`, config);
+        const restaurantResponse = axios.get(`${API_URL}/restaurants/byuser/${user.id}`, config);
```
Comment (ht1204) [inline src/pages/RestaurantAdminView/RestaurantAdminView.jsx:44]: `/restaurants/by-user`

File: src/pages/SignInPage/SignInPage.jsx [hunk 14]
```diff
@@ -226,7 +227,7 @@ const SignInPage = () => {
         } else {
           toast.success(languageSelector(language, 'signInSuccess'));
           setTimeout(() => {
-            navigate('/');
+            navigate(routePaths.home);
           }, 6000);
```
Comment (ht1204) [inline src/pages/SignInPage/SignInPage.jsx:231]: Consider this enhancement:
```js
 const timer = setTimeout(() => {
   navigate(routePaths.home);
 }, 6_000);
clearTimeout(timer);
```

File: src/pages/RestaurantAdminView/RestaurantAdminView.jsx (hunk 13 above)
Comment (CamiloGilFranco) [reply src/pages/RestaurantAdminView/RestaurantAdminView.jsx:44]: This component is being worked on by Nicolas at the moment, so I prefer not to modify his code at this time.

File: src/pages/SignInPage/SignInPage.jsx (hunk 14 above)
Comment (CamiloGilFranco) [reply src/pages/SignInPage/SignInPage.jsx:231]: Same here

File: src/pages/RestaurantAdminView/RestaurantAdminView.jsx (hunk 13 above)
Comment (ht1204) [reply src/pages/RestaurantAdminView/RestaurantAdminView.jsx:44]: ok, no problem, let's move it to approve

**Review by ht1204 — APPROVED** [review]

**Pull request merged**